File: android/app/src/main/java/com/oldsun/classschedule/SummerScheduleMigration.java
package com.oldsun.classschedule;

import android.app.AlarmManager;
import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;
import java.util.List;

final class SummerScheduleMigration {
    private static final String TAG = "SummerScheduleMigration";
    private static final String PREFS_NAME = "CapacitorStorage";
    private static final String SOURCE_VERSION_KEY = "summerScheduleDefaultVersion";
    private static final String SOURCE_KEY = "summerScheduleSource";
    private static final String MIGRATION_VERSION_KEY = "summerScheduleMigrationVersion";
    private static final String MIGRATION_VERSION = "1";
    private static final String SUMMER_THEME_KEY = "summerTheme";
    private static final String THEME_KEY = "theme";

    private static final String[] RESET_KEYS = new String[] {
        "semesterStartDate",
        "notificationsLastScheduledAt",
        "notificationsLastReconciledAt",
        "notificationPlanSnapshot",
        "userGroup",
        "selectedElectives",
        "displayMode",
        "widgetScheduleSnapshot",
        "customSchedule",
        "scheduleSource",
        "remoteScheduleSnapshot",
        "remoteScheduleMeta",
        "remoteSkippedUpdate",
        "defaultScheduleVersion",
        "defaultScheduleSignature",
        "remoteScheduleLastCheckAt",
        "remoteScheduleLastForegroundCheckAt",
        "remoteScheduleLastErrorAt",
        "userExams",
        "summerNotificationsLastScheduledAt",
        "summerNotificationsLastReconciledAt",
        "summerNotificationPlanSnapshot",
        "summerWidgetScheduleSnapshot",
        "summerScheduleCustom",
        "summerScheduleSource",
        "summerScheduleRemoteSnapshot",
        "summerScheduleRemoteMeta",
        "summerScheduleRemoteSkippedUpdate",
        "summerScheduleDefaultSignature",
        "summerScheduleRemoteLastCheckAt",
        "summerScheduleRemoteLastForegroundCheckAt",
        "summerScheduleRemoteLastErrorAt",
        SOURCE_VERSION_KEY,
        SUMMER_THEME_KEY,
        THEME_KEY
    };

    private SummerScheduleMigration() {}

    static boolean ensureMigrated(Context context) {
        if (context == null) return false;

        SharedPreferences preferences =
            context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        if (MIGRATION_VERSION.equals(preferences.getString(MIGRATION_VERSION_KEY, null))) {
            return true;
        }
        if (!preferences.contains(SOURCE_VERSION_KEY) && !preferences.contains(SOURCE_KEY)) {
            return true;
        }

        String summerTheme = preferences.getString(SUMMER_THEME_KEY, null);
        cancelScheduledNotifications(context);

        SharedPreferences.Editor editor = preferences.edit();
        for (String key : RESET_KEYS) {
            editor.remove(key);
        }
        if ("m3".equals(summerTheme) || "minimal".equals(summerTheme)) {
            editor.putString(THEME_KEY, summerTheme);
        }
        editor.putString(MIGRATION_VERSION_KEY, MIGRATION_VERSION);

        boolean migrated = editor.commit();
        if (!migrated) {
            Log.e(TAG, "Failed to persist summer schedule migration");
        }
        return migrated;
    }

    private static void cancelScheduledNotifications(Context context) {
        AlarmClockStorage storage = new AlarmClockStorage(context);
        AlarmManager alarmManager =
            (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        List<Integer> ids = storage.getAllPendingIds();
        if (alarmManager != null) {
            for (int id : ids) {
                AlarmClockSchedulerPlugin.cancelAlarm(context, alarmManager, id);
            }
        }
        storage.clear();
    }
}

File: src/services/app/summerScheduleMigration.js
import * as storage from "../../../storage.js";
import {
  STORAGE_KEYS,
  SUMMER_SCHEDULE_MIGRATION_VERSION,
  SUMMER_STORAGE_KEYS,
  THEMES
} from "../../config/constants.js";

export const CURRENT_SEMESTER_RESET_KEYS = [
  STORAGE_KEYS.SEMESTER_START_DATE,
  STORAGE_KEYS.NOTIFICATIONS_LAST_SCHEDULED_AT,
  STORAGE_KEYS.NOTIFICATIONS_LAST_RECONCILED_AT,
  STORAGE_KEYS.NOTIFICATION_PLAN_SNAPSHOT,
  STORAGE_KEYS.USER_GROUP,
  STORAGE_KEYS.SELECTED_ELECTIVES,
  STORAGE_KEYS.DISPLAY_MODE,
  STORAGE_KEYS.WIDGET_SCHEDULE_SNAPSHOT,
  STORAGE_KEYS.CUSTOM_SCHEDULE,
  STORAGE_KEYS.SCHEDULE_SOURCE,
  STORAGE_KEYS.REMOTE_SCHEDULE_SNAPSHOT,
  STORAGE_KEYS.REMOTE_SCHEDULE_META,
  STORAGE_KEYS.REMOTE_SKIPPED_UPDATE,
  STORAGE_KEYS.DEFAULT_SCHEDULE_VERSION,
  STORAGE_KEYS.DEFAULT_SCHEDULE_SIGNATURE,
  STORAGE_KEYS.REMOTE_LAST_CHECK_AT,
  STORAGE_KEYS.REMOTE_LAST_FOREGROUND_CHECK_AT,
  STORAGE_KEYS.REMOTE_LAST_ERROR_AT,
  STORAGE_KEYS.USER_EXAMS
];

export const SUMMER_SEMESTER_RESET_KEYS = Object.values(
  SUMMER_STORAGE_KEYS
).filter(
  (key) =>
    key !== SUMMER_STORAGE_KEYS.DEFAULT_SCHEDULE_VERSION &&
    key !== SUMMER_STORAGE_KEYS.THEME
);

const isValidTheme = (theme) =>
  theme === THEMES.M3 || theme === THEMES.MINIMAL;

const assertStorageResults = (results) => {
  if (results.some((result) => result === false)) {
    throw new Error("summer-schedule-migration-storage-failed");
  }
};

export const migrateSummerScheduleStorage = async ({
  getItem = storage.getItem,
  setItem = storage.setItem,
  removeItem = storage.removeItem
} = {}) => {
  const [sourceVersion, source, completedVersion, summerTheme] =
    await Promise.all([
      getItem(SUMMER_STORAGE_KEYS.DEFAULT_SCHEDULE_VERSION),
      getItem(SUMMER_STORAGE_KEYS.SCHEDULE_SOURCE),
      getItem(STORAGE_KEYS.SUMMER_SCHEDULE_MIGRATION_VERSION),
      getItem(SUMMER_STORAGE_KEYS.THEME)
    ]);

  if (completedVersion === SUMMER_SCHEDULE_MIGRATION_VERSION) {
    return { status: "already-migrated" };
  }
  if (sourceVersion == null && source == null) {
    return { status: "not-needed" };
  }

  const resetResults = await Promise.all(
    [...CURRENT_SEMESTER_RESET_KEYS, ...SUMMER_SEMESTER_RESET_KEYS].map(
      (key) => removeItem(key)
    )
  );
  assertStorageResults(resetResults);

  const themeResult = isValidTheme(summerTheme)
    ? await setItem(STORAGE_KEYS.THEME, summerTheme)
    : await removeItem(STORAGE_KEYS.THEME);
  assertStorageResults([themeResult]);

  const completedResult = await setItem(
    STORAGE_KEYS.SUMMER_SCHEDULE_MIGRATION_VERSION,
    SUMMER_SCHEDULE_MIGRATION_VERSION
  );
  assertStorageResults([completedResult]);

  await Promise.all([
    removeItem(SUMMER_STORAGE_KEYS.DEFAULT_SCHEDULE_VERSION),
    removeItem(SUMMER_STORAGE_KEYS.THEME)
  ]);

  return {
    status: "migrated",
    theme: isValidTheme(summerTheme) ? summerTheme : null
  };
};
